google-cloud-bigquerystorage/src/main/java/com/google/cloud/bigquery/storage/v1/stub/readrows/BigQueryStorageExceptionFactory.java

```diff
@@ -0,0 +1,172 @@
+import java.util.Map;
+import javax.annotation.Nullable;
+import javax.net.ssl.SSLHandshakeException;
+
+public class BigQueryStorageExceptionFactory {
+
+  public enum ErrorCode {
+    CANCELLED(Status.CANCELLED),
+    UNKNOWN(Status.UNKNOWN),
+    INVALID_ARGUMENT(Status.INVALID_ARGUMENT),
+    DEADLINE_EXCEEDED(Status.DEADLINE_EXCEEDED),
+    NOT_FOUND(Status.NOT_FOUND),
+    ALREADY_EXISTS(Status.ALREADY_EXISTS),
+    PERMISSION_DENIED(Status.PERMISSION_DENIED),
+    UNAUTHENTICATED(Status.UNAUTHENTICATED),
+    RESOURCE_EXHAUSTED(Status.RESOURCE_EXHAUSTED),
+    FAILED_PRECONDITION(Status.FAILED_PRECONDITION),
+    ABORTED(Status.ABORTED),
+    OUT_OF_RANGE(Status.OUT_OF_RANGE),
+    UNIMPLEMENTED(Status.UNIMPLEMENTED),
+    INTERNAL(Status.INTERNAL),
+    UNAVAILABLE(Status.UNAVAILABLE),
+    DATA_LOSS(Status.DATA_LOSS),
+    ;
+
+    private static final Map<Integer, ErrorCode> errorByRpcCode;
+
+    static {
+      ImmutableMap.Builder<Integer, ErrorCode> builder = ImmutableMap.builder();
+      for (ErrorCode errorCode : ErrorCode.values()) {
+        builder.put(errorCode.getCode(), errorCode);
+      }
+      errorByRpcCode = builder.build();
+    }
+
+    private final Status.Code code;
+
+    ErrorCode(Status status) {
+      this.code = status.getCode();
+    }
+
+    int getCode() {
+      return this.code.value();
+    }
+
+    Status getGrpcStatus() {
+      return this.code.toStatus();
+    }
+
+    /** @return the corresponding gRPC status code of this {@link ErrorCode}. */
+    public Status.Code getGrpcStatusCode() {
+      return this.code;
+    }
+
+    /**
+     * Returns the error code represents by {@code name}, or {@code defaultValue} if {@code name}
+     * does not map to a known code.
+     */
+    static ErrorCode valueOf(String name, ErrorCode defaultValue) {
+      try {
+        return ErrorCode.valueOf(name);
+      } catch (IllegalArgumentException e) {
+        return defaultValue;
+      }
+    }
+
+    /**
+     * Returns the error code corresponding to a gRPC status, or {@code UNKNOWN} if not recognized.
+     */
+    public static ErrorCode fromGrpcStatus(Status status) {
+      ErrorCode code = errorByRpcCode.get(status.getCode().value());
+      return code == null ? UNKNOWN : code;
+    }
+
+    /**
+     * Returns the error code corresponding to the given status, or {@code UNKNOWN} if not
+     * recognized.
+     */
+    static ErrorCode fromRpcStatus(com.google.rpc.Status status) {
+      ErrorCode code = errorByRpcCode.get(status.getCode());
+      return code == null ? UNKNOWN : code;
+    }
+  }
+
+  public static BigQueryStorageException newBigQueryStorageException(Throwable throwable) {
+    Status status = Status.fromThrowable(throwable);
+    return new BigQueryStorageException(
+        status.getDescription(),
+        throwable.getCause(),
+        status.getCode().value(),
+        isRetryable(ErrorCode.fromGrpcStatus(status), throwable.getCause()));
```

Comment:
Is this sufficient to cause the client to retry on failure? I don't see a test which verifies the retry behavior.